### ISSUE TITLE
Provide a way to set the working directory for the linters

### ISSUE THREAD

**Issue by dhoogfr** [ansible/vscode-ansible#885]
### Problem

The plugin runs ansible-lint / ansible --syntax-check from a working directory that seems to be set to the workspace root directory.
This can cause issues if for example you have the following directory structure

```
./project
├── code
│   ├── collections
│   ├── files
│   ├── inventories
│   ├── library
│   ├── playbooks
│   └── roles
└── keys
```

With the vscode workspace root directory set to "./project" and you have an ansible.cfg file under the code directory.

In such a situation, the linter will not pickup this ansible.cfg file.
In my case it causes an error with "ansible-playbook --syntax-check" as the vault-password-file variable is not picked up and it fails to decrypt the vault file:

> Unexpected error code 1 from execution of: ansible-playbook --syntax-check -i localhost, code/playbooks/config_sftp_s3fs.ymlansible-lint[internal-error]

### Solution

The Python plugin provides a configuration parameter to set the working directory for the linter.
A similar option for the Ansible plugin to set a (relative) path would solve the above problems

### Alternatives

An alternative would be to have an explicit configuration option to specify the vault password file.
But that would not solve the issue with not reading the ansible.cfg file.

Another alternative could be to instead of using the vscode workspace root as working directory, allow the vscode buildin terminal CWD configuration parameter to override it.
But that could break existing setups

**Comment by cidrblock** [comment]
There is a setting to set the command line paramters for ansible-lint, I'm curious if that could be used to set the target directory for linting as a workspace setting

```
"ansible.validation.lint.arguments": ""
```
It's also in the extentsion's settings:

![image](https://github.com/ansible/vscode-ansible/assets/18386516/e67b6039-1461-4ead-8e71-517770dec1fa)

Let me know if that helps?

TY, brad

**Comment by dhoogfr** [comment]
Hi,

I tried to use that setting with the --project-dir parameter, but it did not work.
From the error message it seems that this parameter is not passed to the "ansible-playbook --syntax-check"  command

![image](https://github.com/ansible/vscode-ansible/assets/2979286/864da183-4d98-41ca-ae17-993cfbaf0956)

![image](https://github.com/ansible/vscode-ansible/assets/2979286/4ba3f275-0ad4-4a0d-9d7e-b58124bb047d)

**Comment by ganeshrn** [comment]
@robpomeroy  The extension does support multi-root workspace so you can try adding `project1` and `project2` folder at the root of the workspace using `Add folder to workspace` option and try setting folder level settings.

<img width="1374" alt="Screenshot 2023-06-06 at 8 11 05 PM" src="https://github.com/ansible/vscode-ansible/assets/5292461/4da71275-cf55-43ae-beeb-50546b176de2">

**Comment by ganeshrn** [comment]
@dhoogfr As mentioned in above comment extension support multi root workspace, so you can add multiple folders at the top level of the workspace.

**Comment by robpomeroy** [comment]
> @robpomeroy The extension does support multi-root workspace so you can try adding `project1` and `project2` folder at the root of the workspace using `Add folder to workspace` option and try setting folder level settings.

Yes, thank you, I had spotted that. 👍 Unfortunately my workspace is large, structured and mature, and I was hoping not to have to rearrange it. It's already multi-root, with Ansible, Docker, Polyglot development, Packer, Terraform and WSL and some others all as separate roots (eight in total). Within the Ansible root, I have 37 separate projects. Ideally I'd prefer a fix that doesn't involve a complete rearrange. 🙂

**Comment by hatifnatt** [comment]
Just bumped into this while setting up `ansible-lint`, I have two Ansible projects with common inventory, with directory structure like this
```
/
├── .ansible-lint
├── code-helper.sh*
├── infra/
│   ├── ansible.cfg
│   ├── .ansible-lint
│   ├── requirements/
│   ├── roles/
│   ├── roles_internal/
│   ├── roles_public/
│   ├── templates/
│   └── myplaybook.yml
├── inventories/
│   ├── group_a/
│   └── group_a/
├── private/
│   ├── ansible.cfg
│   ├── .ansible-lint
│   ├── anotherplaybook.yaml
│   └── roles/
└── .vscode/
    └── settings.json
```
When I'm running `ansible-lint` from inside subproject directory (inside `infra` or `private`) - all fine it reads `ansible.cfg` and takes into account `roles_path` parameter where all roles directories are specified.
But  when VSCode is running `ansible-lint` it always run it with project root as `cwd` therefor `ansible-lint` doesn't read `ansible.cfg` and `.ansible-lint` from sub directories and I'm getting error about 'role not found' i.e.

```
# rewrapped for readability
"message": "the role 'private.role' was not found in
/~snip~/ansible/infra/roles:
/home/hatifnatt/.cache/:
/home/hatifnatt/.ansible/roles:
/usr/share/ansible/roles:
/etc/ansible/roles:
/~snip~/ansible/infra"
```
It's easy to notice that `infra/roles_internal/`, `infra/roles_public/` not included in the search path.

First I'm tried to add
```json
"ansible.validation.lint.arguments": "--project-dir ${PWD}/infra",
```
to `.vscode/settings.json`, but that does not help `ansible-lint` still doesn't read `ansible.cfg`, also in a such way I can only resolve issue with one subproject.

For now I take idea from @erichiller and created helper script `code-helper.sh` located in the project root.
```bash
#!/usr/bin/env bash

SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
export ANSIBLE_ROLES_PATH=./infra/roles/:./infra/roles_internal/:./infra/roles_public/:./private/roles/
code "$SCRIPT_DIR"
```
I'm not using `ANSIBLE_CONFIG` because in my case I have multiple ones and they may be not exactly similar.

From VSCode side of thing it would be nice if environment variables defined for integrated terminal would be passed to command line linters, or alternatively additional configuration section in Ansible extension where environment variables can be defined.

**Comment by PurplProto** [comment]
A workaround I've been using is to update the setting for the `ansible-lint` path to prepend it with `ANSIBLE_CONFIG` env var. This should work for multi-root setups too, since you can set this setting per workspace folder as well.

I've only got one Ansible project in this repo, but it's all in an `ansible` subdir like so:
```
/
├── .venv/
├── ansible/
│   ├── ansible.cfg
│   ├── inventory.py
│   ├── playbooks/
│   ├── roles/
│   └── tasks/
└── other-stuff/
    └── ...
```

In VSCode settings, search for the setting ID `ansible.validation.lint.path` then prepend the existing `ansible-lint` value with an env  var like so:
```shell
ANSIBLE_CONFIG="./ansible/ansible.cfg" ansible-lint
#                   ^--- relative to workspace root (or folder root for multi-root setups)
```
For a visual reference:
<img width="475" height="112" alt="Image" src="https://github.com/user-attachments/assets/058911e9-2b91-46e5-bfa7-5a61bafc34e7" />


Finally, make sure your `ansible.cfg` is set with the correct role paths:
```ini
[defaults]
inventory = ./inventory.py
roles_path = ./roles:~/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles
#                ^--- relative to the path of your ansible.cfg
```

I should add, this would only work on *Nix, not sure on a solution for Windows. It may work with WSL though!
